File: cf1316e.java
import java.util.*;
import java.io.*;
public class Main {
 
	public static void main(String[] args) throws IOException {
		BufferedReader r =
		new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(r.readLine());
 
		int N = Integer.parseInt(st.nextToken());
		int P = Integer.parseInt(st.nextToken());
		int K = Integer.parseInt(st.nextToken());
 
		//Read the audience.
		st = new StringTokenizer(r.readLine());
		pair[] A = new pair[N];
		for (int i = 0; i < N; i++) {
			A[i] = new pair(i, Integer.parseInt(st.nextToken()));
		}
		Arrays.sort(A);
 
		//Read the player skill matrix.
		int[][] skill = new int[N][P];
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(r.readLine());
			for (int j = 0; j < P; j++) {
				skill[i][j] = Integer.parseInt(st.nextToken());
			}
		}
 
		//Initialize dp matrix.
		long[][] dp = new long[N + 1][(1 << P)];
		for (int i = 0; i <= N; i++) {
			Arrays.fill(dp[i], -1);
		}
 
		dp[0][0] = 0;
 
		for (int i = 1; i <= N; i++) {
			int ind = A[i - 1].idx;
			for (int m = 0; m < (1 << P); m++) {
 
				//Count the number of bits in the mask.
				int bits = 0;
				for (int j = 0; j < P; j++) {
					if ((m & (1 << j)) > 0) {
						bits++;
					}
				}
 
				int numAud = i - 1 - bits;
				//If the player is added to the audience.
				if (numAud < K) {
					if (dp[i - 1][m] != -1) {
						dp[i][m] = dp[i - 1][m] + A[i - 1].val;
					}
				} else {
					if (dp[i - 1][m] != -1) {
						dp[i][m] = dp[i - 1][m];
					}
				}
 
				//Add the player to the team.
				for (int j = 0; j < P; j++) {
					if ((m & (1 << j)) > 0 ) {
						if ((dp[i - 1][m ^ (1 << j)]) != -1) {
 
							dp[i][m] = Math.max(dp[i][m],
									dp[i - 1][m ^ (1 << j)] + skill[ind][j]);
						}
					}
				}
			}
		}
 
		//Print the answer.
		System.out.println(dp[N][(1<< P)- 1]);
	}
 
	static class pair implements Comparable<pair> {
		int idx, val;
 
		public pair(int idx, int val) {
			this.idx = idx;
			this.val = val;
		}
 
		public int compareTo(pair other) {
			return other.val - val;
		}
	}
}

File: cf1316e.cpp
#include <bits/stdc++.h>
#include <algorithm>
typedef long long ll;
using namespace std;
#define pii pair<int, int>
#define f first
#define s second
#define mp make_pair
const int MAXN = 1e5 + 3;
const int MAXP = 7;
int N, P, K;
int mod = 1e9 + 7;
ll ind[MAXN];
ll A[MAXN];
ll skill[MAXN][7];

/* DP[i][mask] is the mask strength if we pick players from(0...i).
   The mask is of the player positions.
   If a bit is set, that means this position has been taken,
   if it was not set, then the position is free. */
ll dp[MAXN][(1 << MAXP) + 1]; 

bool sortbysecdesc(const pii& a, const pii& b) {
    return a.s > b.s;
}

/* Because I can't get __builtin_popcount() to work on visual studio */
unsigned int countSetBits(unsigned int n) {
    unsigned int count = 0;
    while (n) {
        count += n & 1;
        n >>= 1;
    }
    return count;
}



int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cin >> N >> P >> K;
   
    memset(dp, -1, sizeof(dp));

    vector<pii> A(N);
    for (int i = 0; i < N; i++) {
        int val; cin >> val;
        A.push_back(mp(i, val));
    }

    /* Skill matrix */
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < P; j++) {
            cin >> skill[i][j];
        }
    }

    /* Sort audience score dsc.*/
    sort(A.begin(), A.end() , sortbysecdesc);

    dp[0][0] = 0;

	for (int i = 1; i <= N; i++) {
		int ind = A[i - 1].f;
		for (int mask = 0; mask < (1 << P); mask++) {
            /* Number of players. */
			int bits = countSetBits(mask);

			int numAud = i - bits - 1;
            /* Try adding it to the audience. */
			if (numAud < K) {
				if (dp[i - 1][mask] != -1) {
					dp[i][mask] = dp[i - 1][mask] + A[i - 1].s;
				}
			}
			else {
				if (dp[i - 1][mask] != -1) {
					dp[i][mask] = dp[i - 1][mask];
				}
			}

			// Try adding the player to the team.
			for (int pos = 0; pos < P; pos++) {
				if ((mask & (1 << pos)) > 0 && (dp[i - 1][mask ^ (1 << pos)]) != -1) {
					dp[i][mask] = max(dp[i][mask], dp[i - 1][mask ^ (1 << pos)] + skill[ind][pos]);
				}
			}
		}
	}
    cout << dp[N][(1 << P) - 1] << "\n";



}


